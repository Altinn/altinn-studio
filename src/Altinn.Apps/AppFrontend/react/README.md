## Running the tests

### Lint checks

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run lint`.

### Unit tests

1. Navigate to the folder `src/Altinn.Apps/AppFrontend/react`.
2. Execute `npm ci`. This step is only nescessary if you have not already done it, or if you change branches.
3. Execute `npm run install-deps`. This step is only nescessary if you have not already done it, or if you change branches.
4. Execute `npm run test`.
