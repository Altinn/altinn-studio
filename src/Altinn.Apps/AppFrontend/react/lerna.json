{
  "command": {
    "bootstrap": {
      "hoist": true
    }
  },
  "packages": [
    "./*"
  ],
  "version": "independent"
}
