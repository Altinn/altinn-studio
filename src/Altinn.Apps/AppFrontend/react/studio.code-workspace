{
	"folders": [
		{
			"path": "."
		}
  ],
  "settings": {
    "typescript.surveys.enabled": false,
    "typescript.tsdk": "node_modules/typescript/lib",
    "eslint.alwaysShowStatus": true,
    "eslint.validate": [
      "javascript",
      "javascriptreact",
      { "language": "typescript", "autoFix": true },
      { "language": "typescriptreact", "autoFix": true}
    ]
  }
}
