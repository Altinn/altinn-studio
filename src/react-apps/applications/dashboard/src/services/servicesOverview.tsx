import { Hidden } from '@material-ui/core';
import { Grid } from '@material-ui/core';
import { withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import withWidth, { isWidthUp } from '@material-ui/core/withWidth';
import classNames from 'classnames';
import * as React from 'react';
import { connect } from 'react-redux';
import AltinnFilterChip from '../../../shared/src/components/AltinnFilterChip';
import AltinnIconButton from '../../../shared/src/components/AltinnIconButton';
import AltinnSearchInput from '../../../shared/src/components/AltinnSearchInput';
import { getLanguageFromKey } from '../../../shared/src/utils/language';
import ServicesCategory from './servicesCategory';

export interface IServicesOverviewComponentProvidedProps {
  classes: any;
  width: any;
}
export interface IServicesOverviewComponentProps extends IServicesOverviewComponentProvidedProps {
  language: any;
  services: any;
  allDistinctOwners: string[];
  selectedOwners: string[];
  currentUserName: string;
}

export interface IServicesOverviewComponentState {
  selectedOwners: string[];
  searchString: string;
}

const styles = {
  mar_top_100: {
    marginTop: '100px',
  },
  mar_top_50: {
    marginTop: '50px',
  },
  mar_bot_50: {
    marginBottom: '50px',
  },
  mar_right_20: {
    marginRight: '20px',
  },
  mar_top_20: {
    marginTop: '20px',
  },
  textToRight: {
    textAlign: 'right' as 'right',
  },
  alignToCenter: {
    textAlign: 'center' as 'center',
  },
  elementToRigth: {
    float: 'right' as 'right',
  },
  textSyle: {
    fontSize: '18px',
    fontWeight: 500,
  }
};
const getListOfDistinctServiceOwners = (services: any, currentUser?: string) => {
  const allDistinctServiceOwners: string[] = [];
  services.map((service: any) => {
    const keyToLookFor = service.owner.full_name || service.owner.login;
    if (allDistinctServiceOwners.indexOf(keyToLookFor) === -1) {
      if (currentUser && currentUser === keyToLookFor) {
        return;
      }
      allDistinctServiceOwners.push(keyToLookFor);
    }
  });
  if (currentUser) allDistinctServiceOwners.unshift(currentUser)
  return allDistinctServiceOwners;
};

const getCurrentUsersName = (user: any) => {
  return user.full_name || user.login;
};

class ServicesOverviewComponent extends React.Component<IServicesOverviewComponentProps, IServicesOverviewComponentState> {
  public static getDerivedStateFromProps(_props: IServicesOverviewComponentProps, _state: IServicesOverviewComponentState) {
    return {
      selectedOwners: _props.selectedOwners,
    };
  }

  public state: IServicesOverviewComponentState = {
    selectedOwners: [],
    searchString: '',
  };

  public searchAndFilterServicesIntoCategoriesCategory(hasWriteRights: any) {
    const filteredServices = this.props.services
      .filter((service: any) => {
        const keyToLookFor = service.owner.full_name || service.owner.login;
        if (service.permissions.push === hasWriteRights && this.state.selectedOwners.indexOf(keyToLookFor) !== -1) {
          return service;
        }
      });

    if (!this.state.searchString) {
      return filteredServices;
    }

    return filteredServices.filter((service: any) =>
      service.name.toLowerCase().includes(this.state.searchString.toLowerCase()) ||
      service.description.toLowerCase().includes(this.state.searchString.toLowerCase()),
    );
  }

  public updateListOfSelectedFilters(key: string) {
    const index = this.state.selectedOwners.indexOf(key);
    this.setState((state: any) => {
      if (index > -1) {
        state.selectedOwners.splice(index, 1);
      } else {
        state.selectedOwners.push(key);
      }
      return {
        selectedOwners: state.selectedOwners,
      };
    });
  }

  public updateSearchSting = (event: any) => {
    this.setState({
      searchString: event.target.value,
    });
  }

  public renderFilters() {
    const { classes } = this.props;
    return (
      <Grid item={true} xl={10} lg={10} md={12}>
        {this.props.allDistinctOwners.map((key: string, index: number) => {
          return (
            <AltinnFilterChip
              key={index}
              className={classNames(
                classes.chip,
                classes.mar_right_20,
                classes.mar_top_20)}
              label={key}
              onclickFunction={this.updateListOfSelectedFilters.bind(this, key)}
              active={this.state.selectedOwners.indexOf(key) !== -1}
            />);
        })}
      </Grid>
    );
  }

  public renderSort() {
    const { classes } = this.props;
    return (
      <Grid item={true} xl={2} lg={2} md={2} sm={2} xs={12}>
        <AltinnFilterChip
          key={getLanguageFromKey('dashboard.sorte_services', this.props.language)}
          className={classNames(
            classes.mar_top_20,
            { [classes.elementToRigth]: isWidthUp('sm', this.props.width) })}
          label={getLanguageFromKey('dashboard.sorte_services', this.props.language)}
          onDeleteFunction={() => { return false }}
          sortIcon={true}
          active={false}
        />
      </Grid>
    );
  }

  public render() {
    const { classes, services } = this.props;
    return (
      <div className={classNames(classes.mar_top_100, classes.mar_bot_50)}>
        <Grid container={true} direction='row'>
          <Grid item={true} xl={8} lg={8} md={8} sm={12} xs={12}>
            <Typography component='h1' variant='h1' gutterBottom={true}>
              {getLanguageFromKey('dashboard.main_header', this.props.language)}
            </Typography>
          </Grid>
          <Grid
            item={true}
            xl={4} lg={4} md={4} sm={12} xs={12}
            className={classNames({ [classes.textToRight]: isWidthUp('md', this.props.width) })}
          >
            <AltinnIconButton
              btnText={getLanguageFromKey('dashboard.new_service', this.props.language)}
              iconClass='ai ai-circle-plus'
            />
          </Grid>
        </Grid>
        <Typography className={classNames(classes.mar_top_50, classes.textSyle)} gutterBottom={true}>
          {getLanguageFromKey('dashboard.main_subheader', this.props.language)}
        </Typography>
        {services &&
          <>
            <Grid container={true} direction='row' className={classes.mar_top_50}>
              <Grid
                item={true}
                xl={12} lg={12} md={10} sm={10} xs={12}
                className={classNames({
                  [classes.alignToCenter]: isWidthUp('lg', this.props.width),
                })}
              >
                <AltinnSearchInput
                  id={'service-search'}
                  placeholder={getLanguageFromKey('dashboard.search_service', this.props.language)}
                  onChangeFunction={this.updateSearchSting}
                />
              </Grid>
              <Hidden lgUp={true}>
                {this.renderSort()}
                {this.renderFilters()}
              </Hidden>
              <Hidden mdDown={true}>
                {this.renderFilters()}
                {this.renderSort()}
              </Hidden>
            </Grid>
            <ServicesCategory
              header={getLanguageFromKey('dashboard.category_service_write', this.props.language)}
              noServicesMessage={getLanguageFromKey('dashboard.no_category_service_write', this.props.language)}
              className={classNames(classes.mar_top_50)}
              categoryRepos={this.searchAndFilterServicesIntoCategoriesCategory(true)}
            />
            <ServicesCategory
              header={getLanguageFromKey('dashboard.category_service_read', this.props.language)}
              noServicesMessage={getLanguageFromKey('dashboard.no_category_service_read', this.props.language)}
              className={classNames(classes.mar_top_100)}
              categoryRepos={this.searchAndFilterServicesIntoCategoriesCategory(false)}
            />
          </>
        }
      </div>
    );
  }
}

const mapStateToProps = (
  state: IDashboardAppState,
  props: IServicesOverviewComponentProvidedProps,
): IServicesOverviewComponentProps => {
  return {
    classes: props.classes,
    width: props.width,
    language: state.language.language,
    services: state.dashboard.services,
    allDistinctOwners: getListOfDistinctServiceOwners(state.dashboard.services, getCurrentUsersName(state.dashboard.user)),
    selectedOwners: getListOfDistinctServiceOwners(state.dashboard.services),
    currentUserName: getCurrentUsersName(state.dashboard.user),
  };
};

export default withWidth()(withStyles(styles)(connect(mapStateToProps)(ServicesOverviewComponent)));