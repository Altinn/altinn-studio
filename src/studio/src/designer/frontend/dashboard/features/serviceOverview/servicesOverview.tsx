/* eslint-disable consistent-return */
/* eslint-disable array-callback-return */
/* eslint-disable no-underscore-dangle */
import { Grid } from '@material-ui/core';
import { createStyles, withStyles } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import withWidth, { isWidthUp } from '@material-ui/core/withWidth';
import classNames from 'classnames';
// import * as DOMPurify from 'dompurify';
// import * as marked from 'marked';
import * as React from 'react';
// import ReactHtmlParser from 'react-html-parser';
import { connect } from 'react-redux';
// import AltinnInformationPaper from 'app-shared/components/AltinnInformationPaper';
// import AltinnLink from 'app-shared/components/AltinnLink';
import AltinnSearchInput from 'app-shared/components/AltinnSearchInput';
import { getLanguageFromKey } from 'app-shared/utils/language';
// import { get } from 'app-shared/utils/networking';
import { CreateNewService } from '../createService/createNewService';
import { ServicesCategory } from './servicesCategory';

export interface IServicesOverviewComponentProvidedProps {
  classes: any;
  width: any;
}
export interface IServicesOverviewComponentProps extends IServicesOverviewComponentProvidedProps {
  language: any;
  services: any[];
  allDistinctOwners: string[];
  selectedOwners: string[];
  currentUserName: string;
}

export interface IServicesOverviewComponentState {
  selectedOwners: string[];
  searchString: string;
  // majorIssues: JSX.Element[];
}

const styles = createStyles({
  mar_top_100: {
    marginTop: '100px',
  },
  mar_top_50: {
    marginTop: '50px',
  },
  mar_bot_50: {
    marginBottom: '50px',
  },
  mar_right_20: {
    marginRight: '20px',
  },
  mar_top_20: {
    marginTop: '20px',
  },
  textToRight: {
    textAlign: 'right' as 'right',
  },
  alignToCenter: {
    textAlign: 'center' as 'center',
  },
  elementToRigth: {
    float: 'right' as 'right',
  },
  textSyle: {
    fontSize: '18px',
    fontWeight: 500,
  },
  paperList: {
    paddingTop: 13,
    paddingBottom: 13,
    fontSize: 16,
  },
  font_16: {
    fontSize: 16,
  },
  mar_top_13: {
    marginTop: 13,
  },
});

const getListOfDistinctServiceOwners = (services: any, currentUser?: string) => {
  const allDistinctServiceOwners: string[] = [];
  if (services) {
    // eslint-disable-next-line array-callback-return
    services.map((service: any) => {
      const keyToLookFor = service.owner.full_name || service.owner.login;
      if (allDistinctServiceOwners.indexOf(keyToLookFor) === -1) {
        if (currentUser === keyToLookFor) {
          return;
        }
        allDistinctServiceOwners.push(keyToLookFor);
      }
    });
  }
  if (currentUser) {
    allDistinctServiceOwners.unshift(currentUser);
  }

  return allDistinctServiceOwners;
};

export const getListOfServicesExcludingCodelist = (services: any) => {
  if (services) {
    return services.filter((service: any) => service.name !== 'codelists');
  }
  return services;
};

const getCurrentUsersName = (user: any) => {
  return user ? user.full_name || user.login : '';
};

// eslint-disable-next-line max-len
export class ServicesOverviewComponent extends React.Component<IServicesOverviewComponentProps, IServicesOverviewComponentState> {
  // eslint-disable-next-line max-len
  public static getDerivedStateFromProps(_props: IServicesOverviewComponentProps) {
    return {
      selectedOwners: _props.selectedOwners,
    };
  }

  public _isMounted = false;

  // eslint-disable-next-line react/state-in-constructor
  public state: IServicesOverviewComponentState = {
    selectedOwners: [],
    searchString: '',
    // majorIssues: null,
  };

  public componentDidMount() {
    this._isMounted = true;
    /* get('https://raw.githubusercontent.com/Altinn/altinn-studio/master/KNOWNISSUES.md')
    .then((res) => {
      if (this._isMounted) {
        marked.setOptions({
          headerIds: false,
        });

        const unsafeHTML = marked(res);
        const safeHTML = DOMPurify.sanitize(unsafeHTML,
          {
            ALLOWED_TAGS: ['ul', 'li', 'a'],
            ALLOWED_ATTR: ['href', 'target'],
          },
        );

        const doc = new DOMParser().parseFromString(safeHTML, 'text/html');
        if (doc.getElementsByTagName('ul').length > 0) {
          const majorWeaknessesList = ReactHtmlParser(doc.getElementsByTagName('ul')[0].innerHTML);
          this.setState({
            majorIssues: majorWeaknessesList,
          });
        }
      }
    }); */
  }

  public componentWillUnmount() {
    this._isMounted = false;
  }

  public updateSearchString = (event: any) => {
    this.setState({
      searchString: event.target.value,
    });
  }

  public searchAndFilterServicesIntoCategoriesCategory(hasWriteRights: any) {
    const filteredServices = this.props.services
      .filter((service: any) => {
        const keyToLookFor = service.owner.full_name || service.owner.login;
        if (service.permissions.push === hasWriteRights && this.state.selectedOwners.indexOf(keyToLookFor) !== -1) {
          return service;
        }
      });

    if (!this.state.searchString) {
      return filteredServices;
    }

    return filteredServices.filter((service: any) => {
      const searchFor = this.state.searchString.toLowerCase();
      if (service.name.toLowerCase().indexOf(searchFor) > -1) return service;
      if (service.owner.login.toLowerCase().indexOf(searchFor) > -1) return service;
      if (service.owner.full_name.toLowerCase().indexOf(searchFor) > -1) return service;
      if (service.description.toLowerCase().indexOf(searchFor) > -1) return service;
    });
  }

  public render() {
    const {
      classes, services, currentUserName,
    } = this.props;
    // const altinnWindow: Window = window;
    // const knownIssuesUrl = `${altinnWindow.location.origin}#/known-issues`;
    return (
      <div className={classNames(classes.mar_top_100, classes.mar_bot_50)}>
        <Grid container={true} direction='row'>
          <Grid
            item={true} xl={8}
            lg={8} md={8}
            sm={12} xs={12}
          >
            <Typography
              component='h1' variant='h1'
              gutterBottom={true}
            >
              {getLanguageFromKey('dashboard.main_header', this.props.language)}
            </Typography>
          </Grid>
          {currentUserName &&
            <Grid
              item={true}
              xl={4}
              lg={4}
              md={4}
              sm={12}
              xs={12}
              className={classNames({ [classes.textToRight]: isWidthUp('md', this.props.width) })}
            >
              <CreateNewService />
            </Grid>
          }
        </Grid>
        {/* this.state.majorIssues &&
          <div className={classes.mar_top_13}>
            <AltinnInformationPaper>
              <Typography className={classes.font_16}>
                {getLanguageFromKey('dashboard.known_issues_subheader', this.props.language)}
              </Typography>

              <Typography className={classes.paperList} component={'ul'}>{this.state.majorIssues}</Typography>

              <AltinnLink
                url={knownIssuesUrl}
                linkTxt={getLanguageFromKey('dashboard.known_issues_link', this.props.language)}
                shouldShowIcon={true}
              />
            </AltinnInformationPaper>
          </div> */
        }
        {/* <Typography className={classNames(classes.mar_top_50, classes.textSyle)} gutterBottom={true}>
          {getLanguageFromKey('dashboard.main_subheader', this.props.language)}
        </Typography> */}
        {services &&
          <>
            <Grid
              container={true} direction='row'
              className={classes.mar_top_13}
            >
              <Grid
                item={true}
                xl={12}
                lg={12}
                md={12}
                sm={12}
                xs={12}
                className={classNames({
                  [classes.alignToCenter]: isWidthUp('md', this.props.width),
                })}
              >
                <AltinnSearchInput
                  id='service-search'
                  ariaLabel={getLanguageFromKey('dashboard.search_service', this.props.language)}
                  placeholder={getLanguageFromKey('dashboard.search_service', this.props.language)}
                  onChangeFunction={this.updateSearchString}
                />
              </Grid>
            </Grid>
            <ServicesCategory
              header={getLanguageFromKey('dashboard.category_service_write', this.props.language)}
              noServicesMessage={getLanguageFromKey('dashboard.no_category_service_write', this.props.language)}
              className={classNames(classes.mar_top_13)}
              categoryRepos={this.searchAndFilterServicesIntoCategoriesCategory(true)}
            />
            <ServicesCategory
              header={getLanguageFromKey('dashboard.category_service_read', this.props.language)}
              noServicesMessage={getLanguageFromKey('dashboard.no_category_service_read', this.props.language)}
              className={classNames(classes.mar_top_50)}
              categoryRepos={this.searchAndFilterServicesIntoCategoriesCategory(false)}
            />
          </>
        }
      </div>
    );
  }
}

const mapStateToProps = (
  state: IDashboardAppState,
  props: IServicesOverviewComponentProvidedProps,
): IServicesOverviewComponentProps => {
  return {
    classes: props.classes,
    width: props.width,
    language: state.language.language,
    services: getListOfServicesExcludingCodelist(state.dashboard.services),
    // eslint-disable-next-line max-len
    allDistinctOwners: getListOfDistinctServiceOwners(state.dashboard.services, getCurrentUsersName(state.dashboard.user)),
    selectedOwners: getListOfDistinctServiceOwners(state.dashboard.services),
    currentUserName: getCurrentUsersName(state.dashboard.user),
  };
};

export const ServicesOverview = withWidth()(withStyles(styles)(connect(mapStateToProps)(ServicesOverviewComponent)));
