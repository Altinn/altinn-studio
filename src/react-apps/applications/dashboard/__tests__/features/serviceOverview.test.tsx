import { mount } from 'enzyme';
import 'jest';
import * as React from 'react';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { ServicesOverview, ServicesOverviewComponent } from '../../src/features/serviceOverview/servicesOverview';

describe('>>> features/serviceOverview', () => {
  let mockStore: any;
  beforeEach(() => {
    const createStore = configureStore();
    const initialState = {
      dashboard: {
        services: [
          {
            name: 'testService',
            owner: { full_name: 'Ulf Utvikler' },
            permissions: true,
          },
          {
            name: 'NullSkatt',
            owner: { full_name: 'Ulf Utvikler' },
            permissions: true,
          },
        ],
      },
      language: {
        language: {
          dashboard: {
            main_header: 'Tjenesteoversikt',
            known_issues_subheader: 'known_issues_subheader',
            known_issues_link: 'known_issues_link',
            main_subheader: 'main_subheader',
          },
        },
      },
    };
    mockStore = createStore(initialState);
  });
  it('>>> Capture snapshot of serviceOverview', () => {
    const rendered = renderer.create(
      <Provider store={mockStore}>
         <ServicesOverview />
      </Provider>,
    );
    expect(rendered).toMatchSnapshot();
  });
  it('+++ should run searchAndFilterServicesIntoCategoriesCategory on search', () => {
    const mountedServiceOverview = mount(
      <Provider store={mockStore}>
         <ServicesOverview />
      </Provider>,
    );
    const instance = mountedServiceOverview.find('ServicesOverviewComponent').instance() as ServicesOverviewComponent;
    const spy = jest.spyOn(instance, 'searchAndFilterServicesIntoCategoriesCategory');

    const searchField = mountedServiceOverview.find('input');
    searchField.simulate('change', { target: {value: 'test'}});
    instance.forceUpdate();
    expect(spy).toHaveBeenCalled();
  });
});
