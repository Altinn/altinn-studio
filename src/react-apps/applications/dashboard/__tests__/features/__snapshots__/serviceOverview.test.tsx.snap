// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> features/serviceOverview >>> Capture snapshot of serviceOverview 1`] = `
<div
  className="Connect-ServicesOverviewComponent--mar_top_100-1 Connect-ServicesOverviewComponent--mar_bot_50-3"
>
  <div
    className="MuiGrid-container-13"
  >
    <div
      className="MuiGrid-item-14 MuiGrid-grid-xs-12-53 MuiGrid-grid-sm-12-67 MuiGrid-grid-md-8-77 MuiGrid-grid-lg-8-91 MuiGrid-grid-xl-8-105"
    >
      <h1
        className="MuiTypography-root-110 MuiTypography-h1-122 MuiTypography-gutterBottom-137"
      >
        Tjenesteoversikt
      </h1>
    </div>
  </div>
  <p
    className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-gutterBottom-137 Connect-ServicesOverviewComponent--mar_top_50-2 Connect-ServicesOverviewComponent--textSyle-9"
  >
    main_subheader
  </p>
  <div
    className="MuiGrid-container-13 Connect-ServicesOverviewComponent--mar_top_50-2"
  >
    <div
      className="MuiGrid-item-14 MuiGrid-grid-xs-12-53 MuiGrid-grid-sm-10-65 MuiGrid-grid-md-10-79 MuiGrid-grid-lg-12-95 MuiGrid-grid-xl-12-109"
    >
      <div
        className="MuiFormControl-root-149 AltinnSearchInput-searchBox-146 MuiFormControl-fullWidth-152"
      >
        <div
          className="MuiFormControl-root-149"
        >
          <div
            className="MuiInputBase-root-166 MuiInput-root-153 AltinnSearchInput-searchBoxInput-147 MuiInputBase-formControl-167 MuiInput-formControl-154 MuiInputBase-adornedStart-170"
            onClick={[Function]}
          >
            <div
              className="MuiInputAdornment-root-183 AltinnSearchInput-searchBoxIcon-148 MuiInputAdornment-positionEnd-186"
            >
              <i
                className="fa fa-search"
              />
            </div>
            <input
              aria-invalid={false}
              className="MuiInputBase-input-176 MuiInput-input-161 MuiInputBase-inputAdornedStart-181"
              disabled={false}
              id="service-search"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="dashboard.search_service"
              required={false}
              type="text"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Connect-ServicesOverviewComponent--mar_top_50-2"
  >
    <div
      className="MuiGrid-container-13"
    >
      <p
        className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-noWrap-136 MuiTypography-gutterBottom-137 CategoryComponent-displayInlineBlock-189 CategoryComponent-width100-190 CategoryComponent-fontSize_16-192"
      >
        dashboard.category_service_write
      </p>
    </div>
    <div
      className="MuiGrid-container-13 MuiGrid-spacing-xs-24-37"
    >
      <div
        className="MuiGrid-item-14 MuiGrid-grid-xs-12-53 MuiGrid-grid-sm-12-67 MuiGrid-grid-md-6-75 MuiGrid-grid-lg-4-87 MuiGrid-grid-xl-3-100 CategoryComponent-width100-190"
      >
        <div
          className="MuiPaper-root-203 MuiPaper-elevation0-205 MuiPaper-rounded-204 MuiCard-root-202 Connect-ServiceCardComponent--card-195"
        >
          <button
            className="MuiButtonBase-root-233 MuiCardActionArea-root-230"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <div
              className="MuiCardContent-root-236"
            >
              <div
                className="MuiGrid-container-13 MuiGrid-spacing-xs-8-35"
              >
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-10-51 MuiGrid-grid-sm-10-65 MuiGrid-grid-md-10-79 MuiGrid-grid-lg-10-93 MuiGrid-grid-xl-10-107"
                >
                  <h3
                    className="MuiTypography-root-110 MuiTypography-h3-124 MuiTypography-noWrap-136 Connect-ServiceCardComponent--displayInlineBlock-193 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--fontSize_16-199 Connect-ServiceCardComponent--fontWeight_500-200"
                  >
                    testService
                  </h3>
                </div>
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-1-42 MuiGrid-grid-sm-1-56 MuiGrid-grid-md-1-70 MuiGrid-grid-lg-1-84 MuiGrid-grid-xl-1-98"
                >
                  <i
                    aria-hidden="true"
                    className="Connect-ServiceCardComponent--iconStyling-196 fa fa-private"
                  />
                </div>
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-1-42 MuiGrid-grid-sm-1-56 MuiGrid-grid-md-1-70 MuiGrid-grid-lg-1-84 MuiGrid-grid-xl-1-98"
                >
                  <i
                    aria-hidden="true"
                    className="Connect-ServiceCardComponent--iconStyling-196 fa fa-write"
                  />
                </div>
                <div
                  className="MuiGrid-item-14 Connect-ServiceCardComponent--displayInlineBlock-193 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--height-198"
                >
                  <p
                    className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-gutterBottom-137 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--fontSize_14-201"
                  >
                    <truncate-markup
                      lines={2}
                    >
                      <span>
                        
                      </span>
                    </truncate-markup>
                  </p>
                </div>
              </div>
              <div
                className="MuiGrid-container-13"
              >
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-6-47 MuiGrid-grid-sm-6-61 MuiGrid-grid-md-6-75 MuiGrid-grid-lg-6-89 MuiGrid-grid-xl-6-103"
                >
                  <p
                    className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-noWrap-136 Connect-ServiceCardComponent--displayInlineBlock-193 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--fontSize_14-201 Connect-ServiceCardComponent--fontWeight_500-200"
                  >
                    Ulf Utvikler
                  </p>
                </div>
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-6-47 MuiGrid-grid-sm-6-61 MuiGrid-grid-md-6-75 MuiGrid-grid-lg-6-89 MuiGrid-grid-xl-6-103"
                >
                  <p
                    className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-noWrap-136 Connect-ServiceCardComponent--displayInlineBlock-193 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--textToRight-197 Connect-ServiceCardComponent--fontSize_14-201 Connect-ServiceCardComponent--fontWeight_500-200"
                  >
                    dashboard.last_changed_service
                     
                    Invalid date
                  </p>
                </div>
              </div>
            </div>
            <span
              className="MuiCardActionArea-focusHighlight-232"
            />
            <span
              className="MuiTouchRipple-root-237"
            />
          </button>
        </div>
      </div>
      <div
        className="MuiGrid-item-14 MuiGrid-grid-xs-12-53 MuiGrid-grid-sm-12-67 MuiGrid-grid-md-6-75 MuiGrid-grid-lg-4-87 MuiGrid-grid-xl-3-100 CategoryComponent-width100-190"
      >
        <div
          className="MuiPaper-root-203 MuiPaper-elevation0-205 MuiPaper-rounded-204 MuiCard-root-202 Connect-ServiceCardComponent--card-195"
        >
          <button
            className="MuiButtonBase-root-233 MuiCardActionArea-root-230"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex="0"
            type="button"
          >
            <div
              className="MuiCardContent-root-236"
            >
              <div
                className="MuiGrid-container-13 MuiGrid-spacing-xs-8-35"
              >
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-10-51 MuiGrid-grid-sm-10-65 MuiGrid-grid-md-10-79 MuiGrid-grid-lg-10-93 MuiGrid-grid-xl-10-107"
                >
                  <h3
                    className="MuiTypography-root-110 MuiTypography-h3-124 MuiTypography-noWrap-136 Connect-ServiceCardComponent--displayInlineBlock-193 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--fontSize_16-199 Connect-ServiceCardComponent--fontWeight_500-200"
                  >
                    NullSkatt
                  </h3>
                </div>
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-1-42 MuiGrid-grid-sm-1-56 MuiGrid-grid-md-1-70 MuiGrid-grid-lg-1-84 MuiGrid-grid-xl-1-98"
                >
                  <i
                    aria-hidden="true"
                    className="Connect-ServiceCardComponent--iconStyling-196 fa fa-private"
                  />
                </div>
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-1-42 MuiGrid-grid-sm-1-56 MuiGrid-grid-md-1-70 MuiGrid-grid-lg-1-84 MuiGrid-grid-xl-1-98"
                >
                  <i
                    aria-hidden="true"
                    className="Connect-ServiceCardComponent--iconStyling-196 fa fa-write"
                  />
                </div>
                <div
                  className="MuiGrid-item-14 Connect-ServiceCardComponent--displayInlineBlock-193 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--height-198"
                >
                  <p
                    className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-gutterBottom-137 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--fontSize_14-201"
                  >
                    <truncate-markup
                      lines={2}
                    >
                      <span>
                        
                      </span>
                    </truncate-markup>
                  </p>
                </div>
              </div>
              <div
                className="MuiGrid-container-13"
              >
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-6-47 MuiGrid-grid-sm-6-61 MuiGrid-grid-md-6-75 MuiGrid-grid-lg-6-89 MuiGrid-grid-xl-6-103"
                >
                  <p
                    className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-noWrap-136 Connect-ServiceCardComponent--displayInlineBlock-193 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--fontSize_14-201 Connect-ServiceCardComponent--fontWeight_500-200"
                  >
                    Ulf Utvikler
                  </p>
                </div>
                <div
                  className="MuiGrid-item-14 MuiGrid-grid-xs-6-47 MuiGrid-grid-sm-6-61 MuiGrid-grid-md-6-75 MuiGrid-grid-lg-6-89 MuiGrid-grid-xl-6-103"
                >
                  <p
                    className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-noWrap-136 Connect-ServiceCardComponent--displayInlineBlock-193 Connect-ServiceCardComponent--width100-194 Connect-ServiceCardComponent--textToRight-197 Connect-ServiceCardComponent--fontSize_14-201 Connect-ServiceCardComponent--fontWeight_500-200"
                  >
                    dashboard.last_changed_service
                     
                    Invalid date
                  </p>
                </div>
              </div>
            </div>
            <span
              className="MuiCardActionArea-focusHighlight-232"
            />
            <span
              className="MuiTouchRipple-root-237"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Connect-ServicesOverviewComponent--mar_top_100-1"
  >
    <div
      className="MuiGrid-container-13"
    >
      <p
        className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-noWrap-136 MuiTypography-gutterBottom-137 CategoryComponent-displayInlineBlock-189 CategoryComponent-width100-190 CategoryComponent-fontSize_16-192"
      >
        dashboard.category_service_read
      </p>
    </div>
    <div
      className="MuiGrid-container-13"
    >
      <p
        className="MuiTypography-root-110 MuiTypography-body1-119 MuiTypography-alignCenter-133 CategoryComponent-width100-190 CategoryComponent-mar_top_100-188 CategoryComponent-fontSize_24-191"
      >
        dashboard.no_category_service_read
      </p>
    </div>
  </div>
</div>
`;
